feat: add API of replace option for copy

liuxh0 opened the pull request:
Add API functionality to specify replace option when copying dataset.

If implementation is okay, I can also add for enq option in the same way.

Related to #579

CForrest97 commented:
Hi @liuxh0, thanks for adding in the replace option! There's a couple of things I would suggest adding to this before it can be PR'd though. 
Could you unit test adding in replace to confirm the `copyExpectStringSpy` receives the option `replace=true`. And add system tests to confirm that setting `replace=true`has the intended affect (i.e without `replace=true` you get an error but when trying to replace a member with `replace=true` it works)
Secondly, if you look at some of the other API methods, the options are passed as a separate parameter. Could you create an `ICopyDataSetOptions` interface which has an optional replace boolean? `IDeleteDatasetOptions` and `Delete.dataSet` should be good examples on how to handle options

zFernand0 commented:
@liuxh0,
I really appreciate you using `Conventional Commit` standards. :+1:

@MarkAckert @jackjia-ibm @MikeBauerCA,
Any ideas how we can make this a standard across Zowe?

liuxh0 commented:
Hi @CForrest97, thanks for your suggestions. I am new to the project.

Now I have moved the option to `ICopyDataSetOptions` and added some unit tests. 

However, `custom_properties.yaml` is missing because I don't have access to any test system. Can you help me to get access to test system? After that I will add system tests.

liuxh0 commented:
@CForrest97 @zFernand0 
I tried to also add some system tests. As I said above, I can't test locally. 
And then I also realized that Jenkins doesn't run system tests. I'm not sure if these tests pass or not. 
Can you help me moving this forward? Thanks!

CForrest97 commented:
@liuxh0 unfortunately getting a test system isn't too straight forward. It may be worth checking out the #zowe-onboarding channel on the open mainframe project's slack workspace. There's people like Joe Winchester who could help out.
In the mean time, I can fork your repo and confirm the system tests work

liuxh0 commented:
@CForrest97 
That would be great. Thanks!

zFernand0 commented on packages/zosfiles/src/api/methods/copy/doc/ICopyDatasetOptions.ts:
```diff
@@ -0,0 +1,21 @@
+/**
+ * This interface defines the options that can be sent into the copy data set function.
+ */
+export interface ICopyDatasetOptions {
```
I believe this Interface could be extended to support all options on the payload.
However there are more changes to the payload on #599.
I'd suggest merging this in, and making all necessary updates on the other PR :)

zFernand0 commented on packages/zosfiles/__tests__/__system__/api/methods/copy/Copy.system.test.ts:
```diff
@@ -216,5 +216,73 @@ describe("Copy", () => {
+                        { replace: true }
+                    );
+                    contents1 = await Get.dataSet(REAL_SESSION, `${fromDataSetName}(${file1})`);
+                    contents2 = await Get.dataSet(REAL_SESSION, `${toDataSetName}(${file2})`);
```
:+1: Thanks for checking the contents of the replaced dataset 👍

zFernand0 commented on packages/zosfiles/__tests__/api/methods/copy/Copy.unit.test.ts:
```diff
@@ -189,6 +189,46 @@ describe("Copy", () => {
+                        { replace: false }
+                    );
+
+                    expect(copyExpectStringSpy).toHaveBeenCalledTimes(1);
```
:+1: cool use of the top level before each

zFernand0 commented:
Please port to `lts-incremental`

zFernand0 commented:
Bypassing vulnerability check.
Please reference https://github.com/zowe/imperative/issues/327

liuxh0 commented:
@zFernand0 Thanks for letting this PR through. 
I have seen there is a PR merging this to `lts-incremental`. (Sorry I saw your comment too late.)
I think now I'll continue with other options mentioned in #579.

zFernand0 commented:
No problem at all, thanks for implementing : )
